### ISSUE TITLE
Role of guidance map

### ISSUE THREAD

**Issue by qitop** [omron-sinicx/neural-astar#2]
Hi, 
I have a some confusions regarding understanding the role of guidance map.

**the paper says**

_Given a problem instance (i.e., an environmental map annotated with start and goal points), the encoder transforms it into a scalar-valued map representation referred to as a guidance map; 
The differentiable A* module then performs a search with the guidance map to output a search history and a resulting path._

Guidance map is U-NET encoder and It is performing segmentation right
1-)What I understand is that initially, on the raw or binary image it outputs the the segmented regions containing blocked and unblocked areas right?

2-)**scalar-valued map representation:** Does it also outputs the A* path when we backpropagate by taking start and goal nodes? and then later on we update these scalar values termed as **guidance cost** using **differentiable A* module** to find the best optimal path as **Figure 2** shows.

**Comment by yonetaniryo** [comment]
Thank you for the question!

- Unet encoder in our approach does not perform segmentation but gives a scaler-valued map that we refer to as a guidance map. As shown below, the final layer is activated by sigmoid. https://github.com/omron-sinicx/neural-astar/blob/e6e626c4d159b0e4c58ee6ad33c7e03db33d72f4/neural_astar/planner/encoder.py#L37 
- Neural A\* then outputs a path by performing differentiable A\* search on the scalar-valued, guidance map (which is called `pred_cost_maps` in our implementation). In other worlds, the encoder itself does not produce the path. https://github.com/omron-sinicx/neural-astar/blob/e6e626c4d159b0e4c58ee6ad33c7e03db33d72f4/neural_astar/planner/astar.py#L68-L69

**Comment by qitop** [comment]
thank you for your response. 
I got your second point.
but what is the scalar valued map? I mean what does it represent because these are being updated during backpropagation.
Also, the start and goal state will be given to encoder as well, as the paper suggest and not the differential A* am I right?

**Comment by yonetaniryo** [comment]
The scalar map (or guidance map in the paper and `pred_cost_maps` in the code) represents the "cost" to include that location as a part of a solution path. Differentiable A\* finds a path under the policy to minimize the sum of those costs. Therefore, lower values in the map mean that you should include those locations in the output path (as they will decrease the total cost) and higher values mean that you should not. We train Neural A* (more precisely the encoder) so that the path found by the differentiable A\* under such policy is as close as the actual shortest paths.

Not only the encoder but differentiable A\* take start/goal information as input. More specifically, the open-list matrix (O in the paper) is initialized as the start map (as the standard A\* algorithm first initializes the open list with a start node.) Also, the goal information is used in Eq (8) of the paper to judge if the differentiable A\* finds a goal or not.

**Comment by qitop** [comment]
I got it. Thanks a bunch 👍
